### PR TITLE
BoneAttachment3D: Add an enum selection that defines the attachment's bone pose target

### LINKED ISSUE #7820: BoneAttachment3D: Add option to use bone transform without override
### Describe the project you are working on

a 3D shooter game with IK

### Describe the problem or limitation you are having in your project

BoneAttachment3D can either override transform or follow bone global transform (with override)
I am not able to get the bone position according to the animation without the IK override, so I can't do a raycast based on the bone position according to the Animation and not after the override has happened using IK

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add Option to BoneAttachment3D (Use bone transform without override)

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Add Option to BoneAttachment3D (Use bone transform no override) (A variable that is available in the Inspector too)

here is my PR for it : 
https://github.com/godotengine/godot/pull/82192

### If this enhancement will not be used often, can it be worked around with a few lines of script?

it can be worked around by getting myself the bone transform from the skeleton using the function get_bone_global_pose_no_override(), but it is not convenient, atleast for me.

### Is there a reason why this should be core and not an add-on in the asset library?

BoneAttachment3D is internal node, and for usability out of the box.

### PR DESCRIPTION
**Pull request by ywmaa**
Closes: https://github.com/godotengine/godot-proposals/issues/7820

### CODE REVIEW

**Comment by smix8** [comment]
Maybe instead of another bool it would be better to add an enum selection that defines if the attachment should target the bone rest, bone pose or bone override transform. Would make it more flexible and cover more uses.

**Comment by ywmaa** [comment]
> Maybe instead of another bool it would be better to add an enum

I guess defaults to bone pose, to not break compatibility with projects ?

**Comment by ywmaa** [comment]
May I ask what is the difference between bone pose and bone pose override ?

I never figured out the difference even when getting them from the skeleton in GD script.

**Comment by ywmaa** [comment]
> add an enum selection that defines if the attachment should target the bone rest, bone pose or bone override transform.

Done, and I tested it, it works for me.

**Comment by ywmaa** [comment]
sorry for multiple commit to just use clang format, it is my first time contributing to the engine, and even though I am trying to follow this : 
https://docs.godotengine.org/en/stable/contributing/development/index.html

but sometimes I forget ;p

I guess it is just better to squash all of this, and use the title as the commit message.

File: doc/classes/BoneAttachment3D.xml [hunk 1]
```diff
@@ -53,5 +53,22 @@
 		<member name="override_pose" type="bool" setter="set_override_pose" getter="get_override_pose" default="false">
 			Whether the BoneAttachment3D node will override the bone pose of the bone it is attached to. When set to [code]true[/code], the BoneAttachment3D node can change the pose of the bone. When set to [code]false[/code], the BoneAttachment3D will always be set to the bone's transform.
 		</member>
+		<member name="bone_pose_mode" type="int" setter="set_bone_pose_mode" getter="get_bone_pose_mode" enum="BoneAttachment3D.BonePoseMode" default="0">
+			Which of the bone poses that BoneAttachment3D node will use, uses [enum BonePoseMode] values. This feature does nothing if [member override_pose] is true.
```
Comment (AThousandShips) [inline doc/classes/BoneAttachment3D.xml:54]:
```suggestion
			Which of the bone poses that BoneAttachment3D node will use, uses [enum BonePoseMode] values. This does nothing if [member override_pose] is [code]true[/code].
```
Use `[code]` for constants, and don't think we use the phrase "feature" like this, could use "This property" instead

File: scene/3d/bone_attachment_3d.cpp [hunk 2]
```diff
@@ -261,6 +261,16 @@ bool BoneAttachment3D::get_override_pose() const {
 	return override_pose;
 }
 
+void BoneAttachment3D::set_bone_pose_mode(BonePoseMode p_pose_mode) {
+	bone_pose_mode = p_pose_mode;
+
```
Comment (AThousandShips) [inline scene/3d/bone_attachment_3d.cpp:266]:
```suggestion
```
Excessive space

File: scene/3d/bone_attachment_3d.cpp [hunk 3]
```diff
@@ -377,4 +422,12 @@ void BoneAttachment3D::_bind_methods() {
...
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "override_pose"), "set_override_pose", "get_override_pose");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "bone_pose_mode", PROPERTY_HINT_ENUM, "Bone Pose,Bone Pose Override,Bone Pose No Override, Bone Rest"), "set_bone_pose_mode", "get_bone_pose_mode");
+
+	BIND_ENUM_CONSTANT(BONE_POSE);
+	BIND_ENUM_CONSTANT(BONE_POSE_OVERRIDE);
+	BIND_ENUM_CONSTANT(BONE_POSE_NO_OVERRIDE);
+	BIND_ENUM_CONSTANT(BONE_REST);
 }
+
+VARIANT_ENUM_CAST(BoneAttachment3D::BonePoseMode);
```
Comment (AThousandShips) [inline scene/3d/bone_attachment_3d.cpp:433]:
```suggestion
VARIANT_ENUM_CAST(BoneAttachment3D::BonePoseMode);

```
Missing trailing line termination

File: scene/3d/bone_attachment_3d.cpp [hunk 4]
```diff
@@ -377,4 +422,12 @@ void BoneAttachment3D::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::STRING_NAME, "bone_name"), "set_bone_name", "get_bone_name");
 	ADD_PROPERTY(PropertyInfo(Variant::INT, "bone_idx"), "set_bone_idx", "get_bone_idx");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "override_pose"), "set_override_pose", "get_override_pose");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "bone_pose_mode", PROPERTY_HINT_ENUM, "Bone Pose,Bone Pose Override,Bone Pose No Override, Bone Rest"), "set_bone_pose_mode", "get_bone_pose_mode");
```
Comment (AThousandShips) [inline scene/3d/bone_attachment_3d.cpp:425]:
```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::INT, "bone_pose_mode", PROPERTY_HINT_ENUM, "Bone Pose,Bone Pose Override,Bone Pose No Override,Bone Rest"), "set_bone_pose_mode", "get_bone_pose_mode");
```
No space between entries

**Comment by ywmaa** [comment]
@AThousandShips done.

thanks for guiding me.

File: doc/classes/BoneAttachment3D.xml [hunk 5]
```diff
@@ -50,8 +50,25 @@
 		<member name="bone_name" type="String" setter="set_bone_name" getter="get_bone_name" default="&quot;&quot;">
 			The name of the attached bone.
 		</member>
+		<member name="bone_pose_mode" type="int" setter="set_bone_pose_mode" getter="get_bone_pose_mode" enum="BoneAttachment3D.BonePoseMode" default="0">
+			Which of the bone poses that BoneAttachment3D node will use, uses [enum BonePoseMode] values. This property does nothing if [member override_pose] is true.
```
Comment (AThousandShips) [inline doc/classes/BoneAttachment3D.xml:54]:
```suggestion
			Which of the bone poses that BoneAttachment3D node will use, uses [enum BonePoseMode] values. This property does nothing if [member override_pose] is [code]true[/code].
```

**Comment by smix8** [comment]
> May I ask what is the difference between bone pose and bone pose override ?

Oversimplified a skeleton bone has 3 stages until the final bone transform.

bone rest - default transform, e.g. no animation, no IK.
bone pose - the transform of the bone after e.g. animation has changed it.
bone override - the transform that "temporary" replaces the pose transform e.g. IK overrides the animation pose.

The bone override transform replaces the bone pose transform by the set override amount on range 0.0 - 1.0.

0.0 override amount -> 100% bone pose transform.
1.0 override amount -> 100% bone override transform.
Inbetween both are mixed with relative override amount to make the final bone transform.


For the BoneAttachment this means we can have 3-4 distinct meaningful transforms to target.
- The bone rest transform relative to skeleton -> get_bone_global_rest().
- The bone pose transform relative to skeleton -> get_bone_global_pose_no_override().
- The bone override transform relative to skeleton -> get_bone_global_pose_override().
- The bone transform relative to skeleton as the result of mixing pose and override by override amount -> get_bone_global_pose().

Please don't ask why this was all named so confusing. This has a gruesome history of contributor frankenstein stitching and also involves episodes of bloody sacrifices on the altar of the backward-compatibility god.

**Comment by AThousandShips** [comment]
The feature hasn't really been considered or decided on so it's not yet time to put a deadline or schedule on it

**Comment by fire** [comment]
@TokageItLab Can you review?

**Comment by TokageItLab** [comment]
I have sent a PR to deprecate bone pose override, which may supersedes this PR, so please check it out.

- https://github.com/godotengine/godot/pull/87888

**Comment by ywmaa** [comment]
Closed this PR, because @TokageItLab 's new system makes this not needed:

https://godotengine.org/article/design-of-the-skeleton-modifier-3d/


I think now I can put the boneattachment before or after the SkeletonModification3D to get a specific bone pose (after then it gets the pose after the modification, before will get the pose before the modification)

**Comment by TokageItLab** [comment]
To be precise, you need to extend Node3D to create your own BoneAttachment3D, but you do not need to prepare it in the core, since you only need to execute the following code on an arbitrary signal if only to get the transform.

```gdscript
@tool

class_name CustomBoneAttachment3D
extends Node3D

@export_enum(" ") var skeleton: NodePath
@export_enum(" ") var bone: String

func _validate_property(property: Dictionary) -> void:
	if property.name == "bone":
		if skeleton:
			property.hint = PROPERTY_HINT_ENUM
			property.hint_string = skeleton.get_concatenated_bone_names()

# Connect to signal SkeletonModifier3D.modification_processed
func _on_skeleton_update() -> void:
	if !skeleton:
		return
	var bone_idx: int = skeleton.find_bone(bone) # You can cache this as needed for performance.
	var global_pose: Transform3D = skeleton.global_transform * skeleton.get_bone_global_pose(bone_idx)
	self.transform = global_pose
```

**Comment by ywmaa** [comment]
> So we make a conclusion that we do not need to have it in the core

so generally BoneAttachment3D will be depcreated, and now we should use our own scripts.

**Comment by TokageItLab** [comment]
To avoid misunderstanding, it is not deprecated.

The built-in BoneAttachment is intended to be considered only for final appearance as a general use case. In other words, the built-in BoneAttachment simply does not care about the pose during modification of SkeletonModifiers.

For example, the most likely use case, attaching an accessory or other item to a specific bone, only needs to consider the final rendering result.

**Comment by ywmaa** [comment]
Ok, I get it now, thanks for the explanation!

**Pull request closed without merging**